### PR TITLE
Make weird name lints trigger behind cfg_attr

### LINKED ISSUE #97094: `unknown_lints` doesn't trigger in all `cfg_attr`
I tried this code:

```rust
#![deny(warnings)]
#![cfg_attr(all(), allow(lint1))]

#[cfg_attr(all(), allow(lint2))]
mod baz {
    #![cfg_attr(all(), allow(lint3))]
}

#[cfg_attr(all(), allow(lint4))]
pub fn test() {}
```
[Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=dee802073df7fd6a2b4183329210bc1d)

I expected attributs in all 4 positions to cause compiler errors, but just the first one (inner attribute on the root module) does. When using `allow` without `cfg_attr` I get all 4 errors. Came across this while debugging why adding `#![cfg_attr(feature = "name", allow(unused_macro_rules))]` didn't introduce new warnings on `stable` in a module.

### Meta

This happens in at least:
- `1.60.0`
- `1.61.0-beta.7`
- `nightly-2022-05-15`

which are the versions currently on the playground

### PR DESCRIPTION
**Pull request by est31**
The weird name lints (`unknown_lints`, `renamed_and_removed_lints`), the lints that lint the linting, were previously not firing for lint level declarations behind `cfg_attr`, as they were only running before expansion.

Now, this will give a `unknown_lints` warning:

```Rust
#[cfg_attr(all(), allow(this_lint_does_not_exist))]
fn foo() {}
```

Lint level declarations behind a `cfg_attr` whose condition is not applying are still ignored. So this still won't give a warning:

```Rust
#[cfg_attr(any(), allow(this_lint_does_not_exist))]
fn foo() {}
```

Furthermore, this PR also makes the weird name lints respect level delcarations for *them* that were hidden by `cfg_attr`, making them consistent to other lints. So this will now not issue a warning:

```Rust
#[cfg_attr(all(), allow(unknown_lints))]
mod foo {
    #[allow(does_not_exist)]
    fn foo() {
    }
}
```

Fixes #97094

### CODE REVIEW

**Comment by lcnr** [comment]
Do we want that, considering that in the future one might write `#[cfg_attr(version = "1.60", allow(some_new_lint)]`? Or am I misunderstanding something here

**Comment by est31** [comment]
@lcnr it will only lint if the condition inside the cfg evaluates to true. If the condition is false, it won't lint. See the last entry in the added test. I've also edited the PR description to make this clearer. Making it peek inside all `cfg_attr` regardless of whether they apply or not is a possibility, but I agree that it's not a good idea, which is why I didn't chose it for this PR.

File: compiler/rustc_lint/src/early.rs [hunk 1]
```diff
@@ -446,12 +446,13 @@ pub fn check_ast_node<'a>(
...
         }
     } else {
+        let pass_count = passes.len();
         for (i, pass) in passes.iter_mut().enumerate() {
             buffered =
                 sess.prof.extra_verbose_generic_activity("run_lint", pass.name()).run(|| {
                     early_lint_node(
                         sess,
-                        pre_expansion && i == 0,
+                        !pre_expansion && i == pass_count,
```
Comment (lcnr) [inline compiler/rustc_lint/src/early.rs:455]: when is `i == pass_count` true? `i` should be in the range `0..pass_count` :thinking:

also, can you add a comment on why that `i == X` check is needed?

File: compiler/rustc_lint/src/early.rs [hunk 2]
```diff
@@ -426,7 +426,7 @@ pub fn check_ast_node<'a>(
     if !sess.opts.debugging_opts.no_interleave_lints {
```
Comment (lcnr) [inline compiler/rustc_lint/src/early.rs:426]: not part of this PR, but can you swap these two branches so that the `if` has a positive condition?

File: compiler/rustc_lint/src/early.rs (hunk 1 above)
Comment (est31) [reply compiler/rustc_lint/src/early.rs:455]: Good catch! The test works great locally but I suppose the other condition of the if is visited.

Comment (est31) [reply compiler/rustc_lint/src/early.rs]: Indeed `i == 0` works too, I've tried it. I've also added a revision to the test file so that it's actually being tested.

**Comment by est31** [comment]
I'm now using `i == 0` because it still works, and have changed the test to ensure that it works with that debug flag enabled. I've also reordered the bodies as requested. re-r? @lcnr

File: compiler/rustc_lint/src/early.rs [hunk 3]
```diff
@@ -424,10 +424,25 @@ pub fn check_ast_node<'a>(
...
     let mut buffered = lint_buffer.unwrap_or_default();
 
-    if !sess.opts.debugging_opts.no_interleave_lints {
+    if sess.opts.debugging_opts.no_interleave_lints {
+        for (i, pass) in passes.iter_mut().enumerate() {
+            buffered =
+                sess.prof.extra_verbose_generic_activity("run_lint", pass.name()).run(|| {
+                    early_lint_node(
+                        sess,
+                        !pre_expansion && i == 0,
```
Comment (lcnr) [inline compiler/rustc_lint/src/early.rs:433]:
```suggestion
                        // We want to only warn about weird lints
                        // once. As it doesn't matter when we lint,
                        // we do it as part of the first pass.
                        !pre_expansion && i == 0,
```
That's my understanding here, is that correct?

**Review by lcnr — COMMENTED** [review]
one nit, then r=me

**Comment by est31** [comment]
@lcnr your understanding is correct, yes. To make sure, I've removed the `== 0` check locally and tested it, and indeed it emitted the warnings multiple times, so the `== 0` is needed. I've added your suggestion to the PR.

File: src/test/ui/lint/issue-97094.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+bar!(
+    #[cfg_attr(all(), allow(nonex_lint_in_macro))]
+    //~^ ERROR unknown lint
+    pub fn _bar() {}
+);
+
+// Allowing unknown lints works if inside cfg_attr
+#[cfg_attr(all(), allow(unknown_lints))]
```
Comment (lcnr) [inline src/test/ui/lint/issue-97094.rs:43]: what happens for
```rust
#[cfg_attr(any(), allow(unknown_lints))]
mod bar {
    #[cfg_attr(all(), allow(nonex_lint_fn))]
    fn _foo() {}
}
``` does that also silence the lint for `nonex_lint_fn`?

Comment (lcnr) [reply src/test/ui/lint/issue-97094.rs:47]: ah damn, that's cool

**Comment by est31** [comment]
Sorry for pinging you again, @lcnr I've improved the last two tests a little to not have a `cfg_attr` inside the module but only a `cfg`. Now they read `#[allow(nonex_lint_fn)]`. I think this improves test readability because the thing being tested in those two last tests is the `cfg_attr` of the module outside. re-r? @lcnr
